File: problems/dp/lc_121-买卖股票的最佳时机.cpp
/**
 *  https://www.bilibili.com/video/BV1ho4y1W7QK/?vd_source=571900c3ae9bbfdc988accacb2feb8be
 */
#include <vector>

class Solution
{
public:
  int maxProfit(std::vector<int> &prices)
  {
    int maxProfit = 0;
    int currMin   = std::numeric_limits<int>::max();
    for (int price : prices) {
      if (price - currMin > maxProfit) {
        maxProfit = price - currMin;
      }
      currMin = std::min(currMin, price);
    }

    return maxProfit;
  }
};

File: problems/dp/lc_121-买卖股票的最佳时机.py
from math import inf
from typing import List

class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        ans = -inf
        currMin = inf
        for price in prices:
            currMin = min(currMin, price)
            ans = max(ans, price - currMin)

        return ans
